GX-TimeFrequency/src/minimax_grids.F90

```diff
@@ -0,0 +1,498 @@
+             mat_A(i_node, j_point) = cos(tau*omega)*2.0_dp*x_mu(i_node)/(x_mu(i_node)**2 + omega**2)
+          end do
+       end do
+
+    ! the sine transform sin(it) -> sin(iw)         
+    else if (transformation_type.eq.3) then
+       omega = omega_points(i_point)
+       current_point = omega
+
+       ! psi(omega_k,x) = 2*omega_k/(x^2+omega_k^2)
+       do i_node = 1, num_x_nodes
+          psi(i_node) = 2.0_dp*omega/((x_mu(i_node))**2 + omega**2)
+       end do
+
+       ! mat_A = sin(omega_k,tau)*phi(tau,x)
+       do j_point = 1, num_points
+          tau = tau_points(j_point)
+          do i_node = 1, num_x_nodes
+             mat_A(i_node, j_point) = sin(omega*tau)*exp(-x_mu(i_node)*tau)
+          end do
+       end do
+    end if
+
+  end subroutine calculate_psi_and_mat_A
+
+  !> /brief calculate the error of the fit function for the cosine/sin transformation for tau to omega or viceversa
+  !! @param[in] num_points: Number of mesh points.
+  !! @param[in] tau_points: imaginary time grid points
+  !! @param[in] omega_points: imaginary frequency grid points
+  !! @param[in] weights_work: work vector for the transformation weights
+  !! @param[in] num_x_nodes: Number of node in the interval [e_min,e_max]
+  !! @param[in] x_mu : Transition energy (nodes in the interval [e_min,e_max])
+  !! @param[in] psi: corresponding auxiliary function (see transformation type definition)
+  !! @param[in] current_point:  current grid point ether omega(i_point) or tau_(i_point)
+  !! @param[out] max_errors: Max error for the three kind of transforms   
+  !! @param[in] transformation type : 1 fit function for the cosine transform cos(it) -> cos(iw), psi(omeaga,x)
+  !!                                : 2 fit function for the cosine transform cos(iw) -> cos(it), psi(tau,x)
+  !!                                : 3 fit function for the sine transform   sin(it) -> sin(iw), psi(omega,x)
+  subroutine calculate_max_error(num_points, tau_points, omega_points, weights_work, num_x_nodes, x_mu, &
+       psi, current_point, max_error, transformation_type)
+
+    real(kind=dp), intent(out)                       :: max_error
+    real(kind=dp), intent(in)                        :: current_point      
+    real(kind=dp), allocatable, dimension(:), &
+         intent(in)                                  :: tau_points, omega_points, x_mu, psi, &
+         weights_work
+    integer, intent(in)                              :: num_points, num_x_nodes
+    integer, intent(in)                              :: transformation_type      
+
+    ! Internal variables
+    integer                                          :: i_node,i_point
+    real(kind=dp)                                    :: func_val, func_val_temp, max_error_tmp, &
+                                                        tau, omega, x_val
+
+    ! Begin work
+    max_error_tmp = 0.0_dp
+
+    ! the cosine transform cos(it) -> cos(iw)
+    if (transformation_type.eq.1) then
+       omega=current_point
+
```

Comment:
Same deal here - this is doing the same thing in each if statement. Turn into a subroutine and pass the appropriate function definition in for` func_val.`